gnomad/utils/constraint.py

```diff
@@ -0,0 +1,162 @@
+    :param use_table_group_by: Whether to force grouping. Defaults to False.
+    :param singleton_expression: Expression for defining a singleton. Defaults to None.
+    :param max_af: Maximum variant AF to keep. By default, no AF cutoff is applied.
+    :param freq_expr: ArrayExpression of Structs with with 'AC' and 'AF' annotations.
+    :param freq_meta_expr: ArrayExpression of meta dictionaries corresponding to freq_expr.
+    :return: Table including 'variant_count' and downsampling counts if requested.
+    """
+    if freq_expr is None and (
+        count_downsamplings or max_af or (count_singletons and singleton_expr is None)
+    ):
+        logger.warning(
+            "freq_expr was not provided, using 'freq' as the frequency annotation."
+        )
+        freq_expr = ht.freq
+    if count_downsamplings and freq_meta_expr is None:
+        logger.warning(
+            "freq_meta_expr was not provided, using 'freq_meta' as the frequency metadata annotation."
+        )
+        freq_meta_expr = ht.freq_meta
+    if count_singletons and singleton_expr is None:
+        logger.warning(
+            "count_singletons is True and singleton_expr was not provided, using freq_expr[0].AC == 1 as the singleton expression."
+        )
+        # Slower, but more flexible (allows for downsampling agg's)
+        singleton_expr = freq_expr[0].AC == 1
+
+    grouping = hl.struct(context=ht.context, ref=ht.ref, alt=ht.alt)
+    if not omit_methylation:
+        grouping = grouping.annotate(methylation_level=ht.methylation_level)
+    for group in additional_grouping:
+        grouping = grouping.annotate(**{group: ht[group]})
+
+    agg = {
+        "variant_count": hl.agg.count_where(freq_expr.AF <= max_af)
+        if max_af
+        else hl.agg.count()
+    }
+    if count_singletons:
+        agg["singleton_count"] = hl.agg.count_where(singleton_expr)
+
+    if count_downsamplings:
+        for pop in count_downsamplings:
+            agg[f"downsampling_counts_{pop}"] = downsampling_counts_expr(
+                freq_expr, freq_meta_expr, pop, max_af=max_af
+            )
+            if count_singletons:
+                agg[f"singleton_downsampling_counts_{pop}"] = downsampling_counts_expr(
+                    freq_expr, freq_meta_expr, pop, singleton=True
+                )
+
+    if use_table_group_by:
+        return ht.group_by(**grouping).partition_hint(partition_hint).aggregate(**agg)
+    else:
+        return ht.aggregate(
+            hl.struct(**{field: hl.agg.group_by(grouping, agg[field]) for field in agg})
+        )
+
+
+def downsampling_counts_expr(
+    freq_expr: hl.expr.ArrayExpression,
+    freq_meta_expr: hl.expr.ArrayExpression,
+    pop: str = "global",
+    variant_quality: str = "adj",
+    singleton: bool = False,
+    max_af: Optional[float] = None,
+) -> hl.expr.ArrayExpression:
+    """
+    Downsample the variant count per given population.
```

Comment:
Indicate what needs to be in the `freq_meta_expr`, how is `variant_quality` and `pop` used?